### PR TITLE
filewatcher: template watcher lazy-loads if enableCodeLenses is off

### PR DESCRIPTION
**Pull request by bryceitoc9**
## Problem
The template file watcher always ran regardless of SAM hint setting

## Solution
The template file watcher now loads on demand. This appears to work even when toggling from off to on (codelenses appear on the template file)
* ~~This may hit the startup function multiple times? Will do some more testing~~ should be resolved with https://github.com/aws/aws-toolkit-vscode/pull/3938/commits/0b1df9d39df91638e749d4825735f548e45ece28
* Confirmed SAM debug works, when coming from startup with codelenses disabled
* Confirmed the registry builds when codelenses are enabled

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/fs/templateRegistry.ts [hunk 1]
```diff
@@ -108,31 +102,37 @@ export class AsyncCloudFormationTemplateRegistry {
...
-                    })
-                    getLogger().debug('cfn: getInstance() requested, still initializing')
-                    while (!this.isSetup) {
-                        await sleep(2000)
+        const config = SamCliSettings.instance
+        if (config.get('enableCodeLenses', false)) {
+            this.asyncSetupFunc(this.instance).then(() => {
+                this.isSetup = true
+                return this.instance
+            })
```
Comment (justinmk3) [inline src/shared/fs/templateRegistry.ts:110]: Isn't the PerfLog still useful to have? It's not only for startup performance, it's for measuring any kind of performance.

File: src/shared/fs/templateRegistry.ts [hunk 2]
```diff
@@ -108,31 +102,37 @@ export class AsyncCloudFormationTemplateRegistry {
...
-                async (progress, token) => {
-                    token.onCancellationRequested(() => {
-                        // Allows for new message to be created if templateRegistry variable attempted to be used again
-                        this.setupProgressMessage = undefined
-                    })
-                    getLogger().debug('cfn: getInstance() requested, still initializing')
-                    while (!this.isSetup) {
-                        await sleep(2000)
+        const config = SamCliSettings.instance
+        if (config.get('enableCodeLenses', false)) {
```
Comment (justinmk3) [inline src/shared/fs/templateRegistry.ts:106]: does sam debugging work with `enableCodeLenses=false` ?

sam debugging depends on the registry to resolve stuff: https://github.com/aws/aws-toolkit-vscode/blob/1db7975286b61ed462e782d7dfc2573a53c615df/src/shared/sam/debugger/awsSamDebugger.ts#L257-L259

Comment (bryceitoc9) [reply src/shared/fs/templateRegistry.ts:107]: Will take a look. If we're getting the instance, it should (it'll build the registry on demand).

File: src/shared/fs/templateRegistry.ts (hunk 1 above)
Comment (bryceitoc9) [reply src/shared/fs/templateRegistry.ts:114]: I'll see if I can worm it in there. It's a little trickier now that this builds on demand and not on the single constructor call.

File: src/shared/fs/templateRegistry.ts (hunk 2 above)
Comment (bryceitoc9) [reply src/shared/fs/templateRegistry.ts:107]: Looks like a no. Will take a look.

Comment (justinmk3) [reply src/shared/fs/templateRegistry.ts:107]: At worst, perhaps remove the `config.get('enableCodeLenses')` check for now. Then we need see what's calling `CloudFormationTemplateRegistry.instance` and try to minimize those calls.

Comment (bryceitoc9) [reply src/shared/fs/templateRegistry.ts]: This triggered a larger, but necessary change: essentially, `globals.templateRegistry` needed to be async to ensure the registry could build.

File: src/shared/cloudformation/activation.ts [hunk 3]
```diff
@@ -79,15 +82,19 @@ function setTemplateRegistryInGlobals(registry: CloudFormationTemplateRegistry)
...
             // already set a mock object before activation.
             // Though in prod nothing should be calling this 'set' function.
             if (this.cfnInstance) {
                 return this.cfnInstance
             }
 
-            return asyncRegistry.getInstance()
+            // prevent eager load if codelenses are off
+            const config = SamCliSettings.instance
+            if (config.get('enableCodeLenses', false) || isToolkitActive()) {
```
Comment (justinmk3) [inline src/shared/cloudformation/activation.ts:95]: Ideally we would fix the callers that are actually calling this during startup. But this is a good compromise for now.

**Review by justinmk3 — APPROVED** [review]
These changes will be helpful, since allowing the global to be async is probably useful. 

But,
- something is still triggering `getInstance()`, even with `aws.samcli.enableCodeLenses = false`. 
- I don't see the "Scanning..." message anymore, even though the logs show that it's scanning.

**Comment by bryceitoc9** [comment]
Going to take some time to figure out why the tests are busted.

> something is still triggering getInstance(), even with aws.samcli.enableCodeLenses = false
On load? This might be the `awsFiletypes` activation. The whole key is that `getInstance` can get hit, but not do anything until it gets hit after activation is finished (so someone booting VS Code and mashing F5 before the activation can load might hit issues)

> I don't see the "Scanning..." message anymore, even though the logs show that it's scanning.
Optimistic me says that this is because it's going so fast that it's missing it (now that it's not necessarily running at the same time as the other filewatcher). I have seen it pop up on my side with the AWS SAM repo.

**Review by justinmk3 — APPROVED** [review]

**Pull request merged**